### PR TITLE
Gradle Version, TargetSdk, CompileSdk Upgraded and VersionCatalogs Implemented

### PR DESCRIPTION
**Pull request by svlilyas**
- Gradle version upgraded to 7.4.2 and Android Gradle Plugin upgraded to 7.2.2
- TargetSdk and CompileSdk upgraded to 32
- buildSrc added to manage Plugins and common Android Configs
- VersionCatalogs implemented with TOML file
- TOML file updated

### CODE REVIEW

**Review by mickael-menu — CHANGES_REQUESTED** [review]
@svlilyas Thank you for taking the time to submit a PR, there's good stuff in it. However, please open an issue (or a Slack message) first next time to discuss the changes, to make sure you don't waste your time on work that won't be accepted.

I left a few comments, but could you also rebase your changes to merge the PR onto the `develop` branch? We only merge hotfixes on `main`, any other changes must be on `develop`.

File: buildSrc/src/main/kotlin/Fields.kt [hunk 1]
```diff
@@ -0,0 +1,5 @@
+object Fields {
+    const val SERVICE_URL = "SERVICE_URL"
+    const val SERVICE_API_KEY = "SERVICE_API_KEY" // optional
+    const val SERVICE_CERTIFICATE_PATH = "SERVICE_CERTIFICATE_PATH" // optional
+}
```
[the remark this answers is not in the document]
Comment (svlilyas) [reply buildSrc/src/main/kotlin/Fields.kt:5]: These used for Flavors based Fields. For example if you want to use SERVICE_URL in UAT Flavor then you need to define there. You can use different values for same variable according to Flavors.

File: gradle.properties [hunk 2]
```diff
@@ -19,3 +19,5 @@ android.useAndroidX=true
 android.enableJetifier=true
 # Kotlin code style for this project: "official" or "obsolete":
 kotlin.code.style=official
+
+android.disableAutomaticComponentCreation=true
```
[the remark this answers is not in the document]
Comment (svlilyas) [reply gradle.properties:23]: It is just for removing the warning about automatic creation to maven publish.

File: buildSrc/src/main/kotlin/ModuleDependency.kt [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    /*
+     Return list of feature modules in the project
+     */
+    fun getFeatureModules(): Set<String> {
+        val featurePrefix = ""
+
+        return getAllModules()
+            .filter { it.startsWith(featurePrefix) }
+            .toSet()
+    }
```
[the remark this answers is not in the document]
Comment (svlilyas) [reply buildSrc/src/main/kotlin/ModuleDependency.kt:32]: You are right, this is just for managing modules. I just use one method here to not hard code module names in Gradle. 
Typing "implementation(data())" is so easy.

File: buildSrc/src/main/kotlin/AndroidConfig.kt [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        const val ENVIRONMENT = "environment"
+    }
+
+    object BuildTypes {
+        const val DEBUG = "debug"
+        const val RELEASE = "release"
+    }
+
+    object Default {
+        const val MAIN = "main"
```
[the remark this answers is not in the document]
Comment (svlilyas) [reply buildSrc/src/main/kotlin/AndroidConfig.kt:40]: As you know Flavors are different from BuildTypes to manage UAT, PILOT, STORE etc. cases.

File: gradle/libs.versions.toml [hunk 5]
```diff
@@ -1,19 +1,17 @@
-# Eventually want to use this for dependencies, but Android Studio is not ready
```
[the remark this answers is not in the document]
Comment (svlilyas) [reply gradle/libs.versions.toml:1]: Nope, unfortunately it is not fully supported to see if there are outdated dependencies :(

File: gradle.properties (hunk 2 above)
[the remark this answers is not in the document]
Comment (mickael-menu) [reply gradle.properties:23]: Alright

File: buildSrc/src/main/kotlin/AndroidConfig.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (mickael-menu) [reply buildSrc/src/main/kotlin/AndroidConfig.kt:40]: Right, but we don't need or use flavors in Readium, so we can remove most of this code.

File: buildSrc/src/main/kotlin/ModuleDependency.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (mickael-menu) [reply buildSrc/src/main/kotlin/ModuleDependency.kt:32]: I'd prefer keeping only the code that is actually used in the build configs.

File: buildSrc/src/main/kotlin/Fields.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (mickael-menu) [reply buildSrc/src/main/kotlin/Fields.kt:5]: We don't use flavors so we can remove the fields.

**Comment by svlilyas** [comment]
I could not able to find rebase settings. Can you tell me how to do it?

**Comment by mickael-menu** [comment]
I changed the base branch of your PR to `develop`, try running `git merge develop` from your clone and fix any conflict you will probably get. This or alternatively you can start again from `develop` with a new PR and copy your changes there.

**Comment by mickael-menu** [comment]
@svlilyas Looks like Android Studio Dolphin has better support for TOML version catalogs 👍

**Pull request closed without merging**